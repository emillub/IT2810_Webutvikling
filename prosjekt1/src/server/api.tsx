const BASE_IMAGE_URL = "https://image.tmdb.org/t/p//w300"
const topRatedMoviesURL = (page : number)=>{return `https://api.themoviedb.org/3/movie/top_rated?language=en-US&page=${page}`}
const GENRE_LIST_URL = "https://api.themoviedb.org/3/genre/movie/list"

const API_OPTIONS = {
    method: 'GET',
    headers: {
      accept: 'application/json',
      Authorization: 'Bearer eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiIyNmZlZDQwNTdlY2U5YjFlMDYwMWE3NzMxZGYyZDYwZSIsIm5iZiI6MTcyNjU4MTA3My44MjA4ODUsInN1YiI6IjY1ZDM3YmZkMTI0MjVjMDE3Y2U1MGJiYyIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.pUtfBzA_GZjIJ-0u0cKXRT3HChmCgumHh2kDdJGR9kA'
    }
  };


export interface movieApiInterface {
    backdrop_path: String,
    id: number,
    title: String,
    original_title: String,
    overview: String,
    poster_path: String,
    adult: boolean,
    original_language: String,
    genre_ids: number[],
    popularity: number,
    release_date: String,
    vote_average: number,
    vote_count: number
}

export interface genreApiInterface{
    id: number,
    name : string
}

interface fetchTopRatedMoviesProps {
    page : number
}

const fetchTopRatedMovies = ({page = 1} : fetchTopRatedMoviesProps) : Promise<{"page" : number, "results": movieApiInterface[]}> => {
    return fetch(topRatedMoviesURL(page), API_OPTIONS).then(res => res.json())
}

const fetchMovieGenres = () : Promise<{"genres" : genreApiInterface}> =>{
    return fetch(GENRE_LIST_URL,API_OPTIONS).then(res => res.json());
}