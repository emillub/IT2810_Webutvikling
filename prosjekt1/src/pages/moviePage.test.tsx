''
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { vi } from 'vitest';
import MoviePage from './moviePage';
import { useTopRatedMovies, fetchMovieGenres } from '../server/api';

vi.mock('../server/api', () => ({
    useTopRatedMovies: vi.fn(),
    fetchMovieGenres: vi.fn(),
    BASE_IMAGE_URL: 'https://image.tmdb.org/t/p/w500',
}));

const queryClient = new QueryClient();

describe('MoviePage', () => {
    beforeEach(() => {
        queryClient.clear();
    });

    it('renders loading state initially', () => {
        (useTopRatedMovies as vi.Mock).mockReturnValue({ data: null, isLoading: true });
        (fetchMovieGenres as vi.Mock).mockResolvedValue({ genres: [] });

        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={['/movie/1']}>
                    <Routes>
                        <Route path="/movie/:movieId" element={<MoviePage />} />
                    </Routes>
                </MemoryRouter>
            </QueryClientProvider>
        );

        expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    it('renders movie details when data is available', async () => {
        const mockMovie = {
            id: 1,
            title: 'Test Movie',
            poster_path: '/test.jpg',
            overview: 'Test overview',
            genre_ids: [1],
            release_date: '2023-01-01',
            original_language: 'en',
        };

        const mockGenres = {
            genres: [{ id: 1, name: 'Action' }],
        };

        (useTopRatedMovies as vi.Mock).mockReturnValue({ data: { results: [mockMovie] }, isLoading: false });
        (fetchMovieGenres as vi.Mock).mockResolvedValue(mockGenres);

        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={['/movie/1']}>
                    <Routes>
                        <Route path="/movie/:movieId" element={<MoviePage />} />
                    </Routes>
                </MemoryRouter>
            </QueryClientProvider>
        );

        expect(await screen.findByText('Test Movie')).toBeInTheDocument();
        expect(screen.getByAltText('Test Movie')).toHaveAttribute('src', 'https://image.tmdb.org/t/p/w500/test.jpg');
        expect(screen.getByText('Test overview')).toBeInTheDocument();
        expect(screen.getByText('Genres: Action')).toBeInTheDocument();
        expect(screen.getByText('Release date: 2023-01-01')).toBeInTheDocument();
        expect(screen.getByText('Original language: en')).toBeInTheDocument();
    });

    it('renders error message when no movie is found', async () => {
        (useTopRatedMovies as vi.Mock).mockReturnValue({ data: { results: [] }, isLoading: false });
        (fetchMovieGenres as vi.Mock).mockResolvedValue({ genres: [] });

        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={['/movie/1']}>
                    <Routes>
                        <Route path="/movie/:movieId" element={<MoviePage />} />
                    </Routes>
                </MemoryRouter>
            </QueryClientProvider>
        );

        expect(await screen.findByText('No movie found')).toBeInTheDocument();
    });
});
