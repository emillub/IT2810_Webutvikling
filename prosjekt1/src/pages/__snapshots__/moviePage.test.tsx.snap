// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MoviePage > renders error message when no movie is found 1`] = `
<div>
  <div
    class="movie-page"
  >
    <p
      class="error-message"
      role="error-message"
    >
      No movie found
    </p>
  </div>
</div>
`;

exports[`MoviePage > renders loading state initially 1`] = `
<div>
  <div
    class="movie-page"
  >
    <p>
      Loading...
    </p>
  </div>
</div>
`;

exports[`MoviePage > renders movie details when data is available 1`] = `
<div>
  <div
    class="movie-page"
  >
    <h1
      class="page-title"
    >
      Test Movie
    </h1>
    <img
      alt="Test Movie"
      class="poster"
      src="https://image.tmdb.org/t/p/w500/test.jpg"
    />
    <section
      class="movieInfo"
    >
      <p>
        Test overview
      </p>
      <p>
        Genres: 
        Action
      </p>
      <p>
        Release date: 
        2023-01-01
      </p>
      <p>
        Original language: 
        en
      </p>
      <button
        class="like-button"
      >
        🤍
      </button>
    </section>
  </div>
</div>
`;
